File: JavaScript/6js 1/bitsWar.js
/*
|--------------------------------------------------------------------------
| World Bits War
|--------------------------------------------------------------------------
|
*/

function bitsWar(numbers) {
   let pOdd = numbers.filter(i => i % 2 != 0 && i > 0).map(j => j.toString(2)).join('').split('').filter(i => i == '1').length;
   let mOdd = numbers.filter(i => i % 2 != 0 && i < 0).map(j => j.toString(2)).join('').split('').filter(i => i == '1').length;
   let pEven = numbers.filter(i => i % 2 == 0 && i > 0).map(j => j.toString(2)).join('').split('').filter(i => i == '1').length;
   let mEven = numbers.filter(i => i % 2 == 0 && i < 0).map(j => j.toString(2)).join('').split('').filter(i => i == '1').length;
   const odd = pOdd - mOdd;
   const even = pEven - mEven;
   return odd > even ? 'odds win' : odd < even ? 'evens win' : 'tie';
}


function bitsWar(num) {
   let [odd, even] = [num.filter(i => i & 1), num.filter(i => !(i & 1))]
   .map(i => i.map(i => i.toString `2`.replace(/0/g, '')
      .replace(/1+/g, i => i.length)).reduce((sum, i) => sum + +i, 0))
   return odd == even ? 'tie' : `${odd > even ? 'odds' : 'evens'} win`
}


function bitsWar(numbers) {
   let hbc = n => [...n.toString(2)].filter(c => c == '1').length;
   let m = 0;
   for (let n of numbers) {
      if (n == 0) continue;
      if (n > 0 ^ n % 2 != 0) m -= hbc(n);
      else m += hbc(n);
   }
   return m == 0 ? "tie" : m > 0 ? "odds win" : "evens win";
}


function bitsWar(numbers) {
   let r = numbers.reduce((a, v) => {
      return a += Math.sign(v) * (v.toString(2).split('1').length - 1) * (v % 2 ? 1 : -1), a
   }, 0)
   return r > 0 ? "odds win" : r < 0 ? "evens win" : "tie"
}


console.log(bitsWar([1, 5, 12]), "odds win");
console.log(bitsWar([7, -3, 20]), "evens win");
console.log(bitsWar([7, -3, -2, 6]), "tie");
console.log(bitsWar([-3, -5]), "evens win");
console.log(bitsWar([]), "tie");

File: Python/6py 1/bits_war.py
"""
|--------------------------------------------------------------------------
| World Bits War
|--------------------------------------------------------------------------
|
"""

def bits_war(numbers):
   pOdd = len(list(filter(lambda i: i == '1', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2!=0 and i>0, numbers))))))))
   mOdd = len(list(filter(lambda i: i == '1', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2!=0 and i<0, numbers))))))))
   pEven = len(list(filter(lambda i: i == '1', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2==0 and i>0, numbers))))))))
   mEven = len(list(filter(lambda i: i == '1', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2==0 and i<0, numbers))))))))
   odd = pOdd-mOdd
   even = pEven-mEven
   return 'odds win' if odd > even else 'evens win' if odd < even else 'tie'


bits_war=lambda n: (lambda s: "evens win" if s>0 else "odds win" if s else "tie")(sum(bin(d)[2:].count("1")*(-1)**((d%2 or d<0)-(d%2 and d<0)) for d in n))


def bits_war(numbers):
   FIGHT = sum( sum(map(int, bin(abs(n))[2:])) * (-1)**(n < 0) * (-1)**(n%2 == 0) for n in numbers )
   return ["evens win", "tie", "odds win"][(FIGHT >= 0) + (FIGHT > 0)]


def bits_war(numbers):
   evens = sum([bin(numb).replace('0b', '').count('1') * (-1 if '-' in bin(numb) else 1) for numb in numbers if numb % 2 == 0])
   odds = sum([bin(numb).replace('0b', '').count('1') * (-1 if '-' in bin(numb) else 1) for numb in numbers if numb % 2 != 0])
   return 'odds win' if odds > evens else 'evens win' if evens > odds else 'tie'


def bits_war(numbers):
   scores = sum(f"{n:b}".count("1") * (-1)**(n%2 == (n>0)) for n in numbers)
   return f"{['odds', 'evens'][scores>0]} win" if scores else "tie"


print(bits_war([1,5,12]), "odds win")
print(bits_war([7,-3,20]), "evens win")
print(bits_war([7,-3,-2,6]), "tie")
print(bits_war([-3,-5]), "evens win")
print(bits_war([]), "tie")